### PR TITLE
Tune multibyte_split kernel

### PR DESCRIPTION
**Pull request by upsj**
## Description
This improves the `multibyte_split` kernel by
* Reducing register pressure: Instead of storing `ITEMS_PER_THREAD` individual states, store only the initial `multistate` for the thread and recompute the individual states on-the-fly
* Eliminating local memory usage: Manipulate the `multistate` via shifts instead of array random access
* Eliminating trie overhead: Since we have only a single delimiter, the trie is a path. We can do the traversal implicitly
* Memoizing which chars were a match: We don't need to recompute this information, but can store it in a bitmask
* Changing the block load algorithm: `BLOCK_LOAD_VECTORIZE` was slightly less efficient than `BLOCK_LOAD_WARP_TRANSPOSE`
* Reducing the allocation overhead by limiting the `output_builder` max allocation size
* Tuning the parameters: `ITEMS_PER_THREAD = 64` works better, and we can improve performance further by operating on larger chunks

Overall, this gives a roughly 2x speedup in my benchmarks.

Based on #11500

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/benchmarks/io/text/multibyte_split.cpp
Comment (bdice) [inline cpp/benchmarks/io/text/multibyte_split.cpp:110]: I wonder if there are ways we could simplify this code by using cudf's [vector factories](https://github.com/rapidsai/cudf/blob/branch-22.10/cpp/include/cudf/detail/utilities/vector_factories.hpp) instead of raw `cudaMemcpyAsync` calls.

File: cpp/benchmarks/io/text/multibyte_split.cpp
Comment (bdice) [inline cpp/benchmarks/io/text/multibyte_split.cpp:124]:
```suggestion
        std::ofstream(temp_file_name, std::ofstream::out).write(host_input.data(), host_input.size());
```

File: cpp/benchmarks/io/text/multibyte_split.cpp
Comment (bdice) [inline cpp/benchmarks/io/text/multibyte_split.cpp:127]: Can we delete these empty comments throughout?
```suggestion
      case data_chunk_source_type::host:
```

**Review by bdice — COMMENTED** [review]
I started reviewing this but realized some of these changes appear to be in https://github.com/rapidsai/cudf/pull/11583? I'm going to wait until that merges to review this.

**Comment by bdice** [comment]
@upsj Ping me once you've had a chance to resolve the conflicts with #11583 and I can re-review. Thanks!

**Comment by upsj** [comment]
@bdice This PR is based on #11500, so you can either take a look at the last few commits only, or wait until that is merged.

File: cpp/src/io/text/multibyte_split.cu
Comment (bdice) [inline cpp/src/io/text/multibyte_split.cu:295]: The `& 1u` seems unnecessary. Should we declare this as `bool const` instead of `auto`, or do a static cast?

**Review by bdice — APPROVED** [review]
These changes look sensible to me. I did not verify the performance improvements.

File: cpp/src/io/text/multibyte_split.cu
Comment (upsj) [reply cpp/src/io/text/multibyte_split.cu:295]: It's necessary to make sure we only read that single bit, otherwise we might also return `true` from more significant bits.

**Comment by upsj** [comment]
I'll post some incremental benchmarks once it's reviewed

File: cpp/src/io/text/multibyte_split.cu
Comment (bdice) [reply cpp/src/io/text/multibyte_split.cu:295]: Sorry, I misinterpreted this on my first glance. It's like `% 2`, not like `> 0`. Resolving.

**Comment by ttnghia** [comment]
I wonder if there is any benchmark showing how much performance improvement this can get?

**Comment by upsj** [comment]
@ttnghia here are the individual changes that had some runtime impact (benchmarked via `benchmarks/MULTIBYTE_SPLIT_BENCHMARK  --axis size_approx[pow2]=30 --axis delim_size=7 --axis delim_percent=1 --axis source_type=0 --axis byte_range_percent=100`)

| commit message | change in runtime|
|---|---|
|improve block loads|-1.04%|
|memoize matches|-14.08%|
|reduce state register pressure|-4.88%|
|store multistate in registers|-39.73%|
|more work per thread|-10.15%|
|bigger chunks|-8.61%|
|tune towards less memory overhead|+2.38%|
|replace trie by device_span|-8.85%|
|**total**|-62.63%|

**Comment by ttnghia** [comment]
Great. I recommend to update that benchmark into the PR description too, so other people can easily see it.

**Review by ttnghia — APPROVED** [review]

**Comment by upsj** [comment]
@gpucibot merge

**Comment by bdice** [comment]
@upsj Amazing work! Thank you for sharing the performance data. https://github.com/rapidsai/cudf/pull/11587#issuecomment-1232974742

**Pull request merged**
